File: src/kernel/fs/gpt.c
/* fs/gpt.c — GUID Partition Table parser */
#include "gpt.h"
#include "drivers/storage/blkdev.h"
#include "lib/klog.h"
#include "lib/string.h"
#include "mm/kmalloc.h"
#include <stdarg.h>

#include <stdint.h>
#include <stdbool.h>

/* ── CRC-32 (IEEE 802.3 polynomial) ─────────────────────────────────────── */
static uint32_t g_crc32_table[256];
static bool     g_crc32_init = false;

static void crc32_init_table(void) {
    for (uint32_t i = 0; i < 256; i++) {
        uint32_t c = i;
        for (int j = 0; j < 8; j++)
            c = (c & 1) ? (0xEDB88320U ^ (c >> 1)) : (c >> 1);
        g_crc32_table[i] = c;
    }
    g_crc32_init = true;
}

static uint32_t crc32(const uint8_t *data, size_t len) {
    if (!g_crc32_init) crc32_init_table();
    uint32_t c = 0xFFFFFFFFU;
    for (size_t i = 0; i < len; i++)
        c = g_crc32_table[(c ^ data[i]) & 0xFF] ^ (c >> 8);
    return c ^ 0xFFFFFFFFU;
}

/* ── GPT on-disk structures ──────────────────────────────────────────────── */
#define GPT_HEADER_SIGNATURE  0x5452415020494645ULL  /* "EFI PART" LE       */
#define GPT_HEADER_REVISION   0x00010000U

typedef struct __attribute__((packed)) {
    uint64_t signature;        /* "EFI PART"                                 */
    uint32_t revision;
    uint32_t header_size;      /* usually 92                                 */
    uint32_t header_crc32;     /* CRC of this header (field zeroed)          */
    uint32_t reserved;
    uint64_t my_lba;           /* LBA of this header (1 for primary)         */
    uint64_t alternate_lba;    /* LBA of backup header                       */
    uint64_t first_usable_lba;
    uint64_t last_usable_lba;
    gpt_guid_t disk_guid;
    uint64_t part_entry_start_lba;
    uint32_t num_part_entries;
    uint32_t part_entry_size;  /* usually 128                                */
    uint32_t part_array_crc32;
    /* rest padded to sector */
} gpt_header_t;

typedef struct __attribute__((packed)) {
    gpt_guid_t type_guid;
    gpt_guid_t part_guid;
    uint64_t   lba_start;
    uint64_t   lba_end;
    uint64_t   flags;
    uint16_t   name[36];   /* UTF-16LE                                       */
} gpt_entry_t;

/* ── Well-known GUIDs ────────────────────────────────────────────────────── */
const gpt_guid_t GPT_GUID_EMPTY = {
    0x00000000, 0x0000, 0x0000,
    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }
};
const gpt_guid_t GPT_GUID_EFI_SYSTEM = {
    0xC12A7328, 0xF81F, 0x11D2,
    { 0xBA, 0x4B, 0x00, 0xA0, 0xC9, 0x3E, 0xC9, 0x3B }
};
const gpt_guid_t GPT_GUID_MS_BASIC_DATA = {
    0xEBD0A0A2, 0xB9E5, 0x4433,
    { 0x87, 0xC0, 0x68, 0xB6, 0xB7, 0x26, 0x99, 0xC7 }
};
const gpt_guid_t GPT_GUID_LINUX_DATA = {
    0x0FC63DAF, 0x8483, 0x4772,
    { 0x8E, 0x79, 0x3D, 0x69, 0xD8, 0x47, 0x7D, 0xE4 }
};
const gpt_guid_t GPT_GUID_LINUX_SWAP = {
    0x0657FD6D, 0xA4AB, 0x43C4,
    { 0x84, 0xE5, 0x09, 0x33, 0xC8, 0x4B, 0x4F, 0x4F }
};

/* ── Helpers ─────────────────────────────────────────────────────────────── */
bool gpt_guid_equal(const gpt_guid_t *a, const gpt_guid_t *b) {
    return memcmp(a, b, sizeof(gpt_guid_t)) == 0;
}

const char *gpt_guid_type_str(const gpt_guid_t *g) {
    if (gpt_guid_equal(g, &GPT_GUID_EFI_SYSTEM))    return "EFI System";
    if (gpt_guid_equal(g, &GPT_GUID_MS_BASIC_DATA)) return "MS Basic Data";
    if (gpt_guid_equal(g, &GPT_GUID_LINUX_DATA))    return "Linux Data";
    if (gpt_guid_equal(g, &GPT_GUID_LINUX_SWAP))    return "Linux Swap";
    if (gpt_guid_equal(g, &GPT_GUID_EMPTY))         return "Empty";
    return "Unknown";
}

/* UTF-16LE → ASCII (lossy: non-ASCII replaced with '?') */
static void utf16le_to_ascii(const uint16_t *src, int max_chars, char *dst) {
    for (int i = 0; i < max_chars; i++) {
        uint16_t c = src[i];
        if (c == 0) { dst[i] = '\0'; return; }
        dst[i] = (c < 0x80) ? (char)c : '?';
    }
    dst[max_chars] = '\0';
}

/* ── Public API ──────────────────────────────────────────────────────────── */
int gpt_scan(blkdev_t *dev, gpt_partition_t *out, int max) {
    if (!dev || !out || max <= 0) return 0;

    /* ── Read LBA 1 (primary GPT header) ── */
    uint8_t sector[512];
    if (blkdev_read(dev, 1, 1, sector) < 0) {
        KLOG_WARN("gpt: '%s': failed to read LBA 1\n", dev->name);
        return 0;
    }

    gpt_header_t hdr;
    memcpy(&hdr, sector, sizeof(hdr));

    /* Validate signature */
    if (hdr.signature != GPT_HEADER_SIGNATURE) {
        KLOG_INFO("gpt: '%s': no GPT signature (not a GPT disk)\n", dev->name);
        return 0;
    }
    if (hdr.revision != GPT_HEADER_REVISION) {
        KLOG_WARN("gpt: '%s': unknown revision 0x%08X\n",
                  dev->name, hdr.revision);
    }

    /* Validate header CRC */
    uint32_t saved_crc = hdr.header_crc32;
    hdr.header_crc32   = 0;
    uint32_t computed  = crc32((uint8_t *)&hdr, hdr.header_size);
    if (computed != saved_crc) {
        KLOG_WARN("gpt: '%s': header CRC mismatch (got %08X expected %08X)\n",
                  dev->name, computed, saved_crc);
        return 0;
    }

    uint32_t nparts  = hdr.num_part_entries;
    uint32_t entry_sz = hdr.part_entry_size;
    uint64_t arr_lba  = hdr.part_entry_start_lba;

    KLOG_INFO("gpt: '%s': %u partition entries, entry_size=%u, arr_lba=%llu\n",
              dev->name, nparts, entry_sz,
              (unsigned long long)arr_lba);

    /* Validate entry size */
    if (entry_sz < sizeof(gpt_entry_t) || entry_sz > 512) {
        KLOG_WARN("gpt: '%s': unexpected entry size %u\n", dev->name, entry_sz);
        return 0;
    }

    /* How many entries fit in one sector */
    uint32_t eps = 512 / entry_sz;  /* entries per sector */
    int count = 0;

    /* Determine parent device name length for partition naming */
    int pname_len = (int)strlen(dev->name);
    /* If parent ends in a digit (e.g. "nvme0n1"), add 'p' separator */
    bool needs_p = (pname_len > 0 &&
                    dev->name[pname_len - 1] >= '0' &&
                    dev->name[pname_len - 1] <= '9');
    int part_num = 1;

    for (uint32_t i = 0; i < nparts && count < max; ) {
        uint64_t lba = arr_lba + i / eps;
        uint8_t  buf[512];
        if (blkdev_read(dev, lba, 1, buf) < 0) break;

        for (uint32_t j = 0; j < eps && i < nparts && count < max; j++, i++) {
            gpt_entry_t *e = (gpt_entry_t *)(buf + j * entry_sz);

            /* Skip empty entries */
            if (gpt_guid_equal(&e->type_guid, &GPT_GUID_EMPTY)) continue;
            if (e->lba_start == 0 && e->lba_end == 0) continue;

            gpt_partition_t *p = &out[count];
            p->type_guid = e->type_guid;
            p->part_guid = e->part_guid;
            p->lba_start = e->lba_start;
            p->lba_end   = e->lba_end;
            p->flags     = e->flags;
            memcpy(p->label, e->name, sizeof(p->label));
            utf16le_to_ascii(e->name, 36, p->label_ascii);

            /* Create partition blkdev name: e.g. "vda1", "sda1", "nvme0n1p1" */
            char part_name[40];
            if (needs_p)
                ksnprintf(part_name, sizeof(part_name), "%sp%d",
                          dev->name, part_num);
            else
                ksnprintf(part_name, sizeof(part_name), "%s%d",
                          dev->name, part_num);

            p->blkdev = blkdev_partition_create(dev, part_name,
                                                 e->lba_start, e->lba_end);

            KLOG_INFO("gpt: '%s' part%d '%s' [%llu..%llu] %s\n",
                      dev->name, part_num, p->label_ascii,
                      (unsigned long long)e->lba_start,
                      (unsigned long long)e->lba_end,
                      gpt_guid_type_str(&e->type_guid));

            count++;
            part_num++;
        }
    }

    KLOG_INFO("gpt: '%s': found %d valid partition(s)\n", dev->name, count);
    return count;
}

File: src/zig/src/gpt.zig
//! src/gpt.zig — GPT partition table parser (Zig rewrite of fs/gpt.c)
//!
//! Implements the same C ABI as fs/gpt.h so it links drop-in.
//! Rules:
//!   * No stdlib — freestanding kernel code.
//!   * All C-visible symbols use `export`.
//!   * No heap allocation; the caller supplies the output array.

const std = @import("std");

// ── Extern C kernel functions ────────────────────────────────────────────────

/// Opaque handle to a block device (blkdev_t * in C).
pub const BlkDev = opaque {};

// blkdev_read(dev, lba, count, buf) → 0 on success, -1 on error
extern fn blkdev_read(dev: *BlkDev, lba: u64, count: u32, buf: [*]u8) c_int;

// blkdev_partition_create(parent, name, lba_start, lba_end) → blkdev_t* or NULL
// NOTE: lba_end is INCLUSIVE — matches the C signature in blkdev.h
extern fn blkdev_partition_create(
    parent:    *BlkDev,
    name:      [*:0]const u8,
    lba_start: u64,
    lba_end:   u64,
) ?*BlkDev;

// Return the ASCII name of a blkdev (the `name[32]` field).
// We declare this as an extern instead of dereferencing the opaque type.
extern fn blkdev_get_name(dev: *BlkDev) [*:0]const u8;

// Kernel logger shim — level 2 = INFO, 1 = WARN
extern fn klog_write_str(level: c_int, msg: [*:0]const u8) void;

// ── Sector I/O ───────────────────────────────────────────────────────────────
const SECTOR: usize = 512;

fn read_lba(dev: *BlkDev, lba: u64, buf: *[SECTOR]u8) bool {
    return blkdev_read(dev, lba, 1, buf) == 0;
}

// ── On-disk GPT structures (packed, little-endian host byte order) ───────────

pub const Guid = extern struct {
    data1: u32,
    data2: u16,
    data3: u16,
    data4: [8]u8,

    pub fn eql(a: *const Guid, b: *const Guid) bool {
        return a.data1 == b.data1 and
               a.data2 == b.data2 and
               a.data3 == b.data3 and
               std.mem.eql(u8, &a.data4, &b.data4);
    }

    pub fn is_zero(g: *const Guid) bool {
        return g.data1 == 0 and g.data2 == 0 and g.data3 == 0 and
               std.mem.allEqual(u8, &g.data4, 0);
    }
};
comptime { std.debug.assert(@sizeOf(Guid) == 16); }

/// GPT header at LBA 1 (first 92 bytes; rest of sector is padding).
const GptHeader = extern struct {
    signature:        [8]u8,    // "EFI PART"
    revision:         u32,
    header_size:      u32,
    header_crc32:     u32,
    _reserved:        u32,
    current_lba:      u64,
    backup_lba:       u64,
    first_usable:     u64,
    last_usable:      u64,
    disk_guid:        Guid,
    part_entry_lba:   u64,
    num_entries:      u32,
    entry_size:       u32,
    entries_crc32:    u32,

    fn valid(h: *const GptHeader) bool {
        return std.mem.eql(u8, &h.signature, "EFI PART");
    }
};

/// One 128-byte GPT partition entry.
const GptEntry = extern struct {
    type_guid:  Guid,
    part_guid:  Guid,
    start_lba:  u64,
    end_lba:    u64,
    attributes: u64,
    name:       [36]u16,  // UTF-16LE partition label
};
comptime { std.debug.assert(@sizeOf(GptEntry) == 128); }

// ── C-facing output struct — must match gpt_partition_t in gpt.h exactly ─────
pub const GptPartition = extern struct {
    type_guid:   Guid,
    part_guid:   Guid,
    lba_start:   u64,
    lba_end:     u64,
    flags:       u64,
    label:       [36]u16,
    label_ascii: [37]u8,
    blkdev:      ?*BlkDev,
};

// ── Well-known type GUIDs ────────────────────────────────────────────────────
export const GPT_GUID_EMPTY = Guid{
    .data1 = 0, .data2 = 0, .data3 = 0, .data4 = .{0} ** 8,
};
export const GPT_GUID_EFI_SYSTEM = Guid{
    .data1 = 0xC12A7328, .data2 = 0xF81F, .data3 = 0x11D2,
    .data4 = .{ 0xBA, 0x4B, 0x00, 0xA0, 0xC9, 0x3E, 0xC9, 0x3B },
};
export const GPT_GUID_MS_BASIC_DATA = Guid{
    .data1 = 0xEBD0A0A2, .data2 = 0xB9E5, .data3 = 0x4433,
    .data4 = .{ 0x87, 0xC0, 0x68, 0xB6, 0xB7, 0x26, 0x99, 0xC7 },
};
export const GPT_GUID_LINUX_DATA = Guid{
    .data1 = 0x0FC63DAF, .data2 = 0x8483, .data3 = 0x4772,
    .data4 = .{ 0x8E, 0x79, 0x3D, 0x69, 0xD8, 0x47, 0x7D, 0xE4 },
};
export const GPT_GUID_LINUX_SWAP = Guid{
    .data1 = 0x0657FD6D, .data2 = 0xA4AB, .data3 = 0x43C4,
    .data4 = .{ 0x84, 0xE5, 0x09, 0x33, 0xC8, 0x4B, 0x4F, 0x4F },
};

// ── Helpers ──────────────────────────────────────────────────────────────────

/// Lossy UTF-16LE → ASCII: replace non-ASCII code units with '?'.
fn utf16le_to_ascii(src: []const u16, dst: []u8) void {
    var i: usize = 0;
    while (i < src.len and i < dst.len - 1) : (i += 1) {
        const cp = src[i];
        if (cp == 0) break;
        dst[i] = if (cp < 0x80) @truncate(cp) else '?';
    }
    dst[i] = 0;
}

/// Build "parentN" device name into `buf`. Returns NUL-terminated slice.
fn fmt_part_name(parent_name: [*:0]const u8, idx: usize, buf: []u8) [:0]u8 {
    var pos: usize = 0;
    // Copy parent name
    var p: usize = 0;
    while (parent_name[p] != 0 and pos < buf.len - 4) : (p += 1) {
        buf[pos] = parent_name[p];
        pos += 1;
    }
    // Append 1-based decimal index
    var n: usize = idx + 1;
    var tmp: [8]u8 = undefined;
    var d: usize = 0;
    if (n == 0) { tmp[d] = '0'; d += 1; }
    while (n > 0) : (n /= 10) {
        tmp[d] = '0' + @as(u8, @truncate(n % 10));
        d += 1;
    }
    while (d > 0) {
        d -= 1;
        if (pos < buf.len - 1) { buf[pos] = tmp[d]; pos += 1; }
    }
    buf[pos] = 0;
    return buf[0..pos :0];
}

// ── Exported C ABI ───────────────────────────────────────────────────────────

/// bool gpt_guid_equal(const gpt_guid_t *a, const gpt_guid_t *b)
export fn gpt_guid_equal(a: *const Guid, b: *const Guid) bool {
    return a.eql(b);
}

/// const char *gpt_guid_type_str(const gpt_guid_t *g)
export fn gpt_guid_type_str(g: *const Guid) [*:0]const u8 {
    if (g.eql(&GPT_GUID_EFI_SYSTEM))    return "EFI System";
    if (g.eql(&GPT_GUID_MS_BASIC_DATA)) return "Microsoft Basic Data";
    if (g.eql(&GPT_GUID_LINUX_DATA))    return "Linux Filesystem";
    if (g.eql(&GPT_GUID_LINUX_SWAP))    return "Linux Swap";
    if (g.is_zero())                     return "Empty";
    return "Unknown";
}

/// int gpt_scan(blkdev_t *dev, gpt_partition_t *out, int max)
///
/// Reads LBA 1 for the GPT header, iterates partition entries, populates
/// `out[]`, and calls blkdev_partition_create() for each valid entry.
/// Returns the count of partitions found (0 on error / empty disk).
export fn gpt_scan(dev: *BlkDev, out: [*]GptPartition, max: c_int) c_int {
    if (max <= 0) return 0;

    // ── Read GPT header at LBA 1 ─────────────────────────────────────────
    var hdr_sector: [SECTOR]u8 align(8) = undefined;
    if (!read_lba(dev, 1, &hdr_sector)) {
        klog_write_str(1, "gpt(zig): failed to read LBA 1\n");
        return 0;
    }

    const hdr = @as(*const GptHeader, @ptrCast(@alignCast(&hdr_sector)));
    if (!hdr.valid()) {
        klog_write_str(1, "gpt(zig): invalid GPT signature\n");
        return 0;
    }

    const entry_size = hdr.entry_size;
    if (entry_size == 0 or entry_size > SECTOR) {
        klog_write_str(1, "gpt(zig): unsupported partition entry size\n");
        return 0;
    }

    const n_want:           u32 = hdr.num_entries;
    const entries_per_sec:  u32 = @intCast(SECTOR / entry_size);
    const part_lba:         u64 = hdr.part_entry_lba;
    const parent_name:  [*:0]const u8 = blkdev_get_name(dev);

    // ── Iterate entries ──────────────────────────────────────────────────
    var found: c_int = 0;
    var name_buf: [32]u8 = undefined;

    var ei: u32 = 0;
    while (ei < n_want and found < max) : (ei += 1) {
        const sec_idx:  u64 = ei / entries_per_sec;
        const slot:     u32 = ei % entries_per_sec;

        var ent_buf: [SECTOR]u8 align(8) = undefined;
        if (!read_lba(dev, part_lba + sec_idx, &ent_buf)) break;

        const byte_off: usize = @as(usize, slot) * entry_size;
        // Safety: ent_buf is align(8); GptEntry is 128 bytes; entry_size >= 128
        const ent = @as(*const GptEntry, @ptrCast(@alignCast(&ent_buf[byte_off])));

        // Skip empty entries
        if (ent.type_guid.is_zero()) continue;
        if (ent.start_lba == 0 or ent.end_lba < ent.start_lba) continue;

        // Fill output
        const slot_out = &out[@intCast(found)];
        slot_out.type_guid = ent.type_guid;
        slot_out.part_guid = ent.part_guid;
        slot_out.lba_start = ent.start_lba;
        slot_out.lba_end   = ent.end_lba;
        slot_out.flags     = ent.attributes;
        slot_out.label     = ent.name;
        utf16le_to_ascii(&ent.name, &slot_out.label_ascii);
        slot_out.blkdev    = null;

        // Register partition blkdev
        const pname = fmt_part_name(parent_name, @intCast(found), &name_buf);
        slot_out.blkdev = blkdev_partition_create(
            dev, pname.ptr, ent.start_lba, ent.end_lba,
        );

        found += 1;
    }

    return found;
}

// ── Panic handler (mandatory for freestanding Zig) ───────────────────────────
pub fn panic(_: []const u8, _: ?*std.builtin.StackTrace, _: ?usize) noreturn {
    while (true) {
        asm volatile ("cli");
        asm volatile ("hlt");
    }
}
